crates/apollo-compiler/src/validation/selection.rs

```diff
@@ -1,112 +1,229 @@
+                        detect_fragment_cycle_inner(
+                            fragments,
+                            &fragment.selection_set,
+                            &mut visited.push(&fragment.name)?,
+                        )?;
+                    }
+                }
+                executable::Selection::InlineFragment(inline) => {
+                    detect_fragment_cycle_inner(fragments, &inline.selection_set, visited)?;
+                }
+                executable::Selection::Field(field) => {
+                    detect_fragment_cycle_inner(fragments, &field.selection_set, visited)?;
                 }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Check if two field selections from the overlapping types are the same, so the fields can be merged.
+fn same_name_and_arguments(
+    field_a: FieldSelection<'_>,
+    field_b: FieldSelection<'_>,
+) -> Result<(), BuildError> {
+    // 2bi. fieldA and fieldB must have identical field names.
+    if field_a.field.name != field_b.field.name {
+        return Err(BuildError::ConflictingFieldName {
+            alias: field_a.field.response_key().clone(),
+            original_location: field_a.field.location(),
+            original_selection: field_a.coordinate(),
+            conflicting_location: field_b.field.location(),
+            conflicting_selection: field_b.coordinate(),
+        });
+    }
+
+    // 2bii. fieldA and fieldB must have identical sets of arguments.
+    let conflicting_field_argument =
+        |original_arg: Option<&Node<ast::Argument>>,
+         redefined_arg: Option<&Node<ast::Argument>>| {
+            debug_assert!(
+                    original_arg.is_some() || redefined_arg.is_some(),
+                    "a conflicting field argument error can only exist when at least one field has the argument",
+                );
+
+            // We can take the name from either one of the arguments as they are necessarily the same.
+            let arg = original_arg.or(redefined_arg).unwrap();
+
+            BuildError::ConflictingFieldArgument {
+                // field_a and field_b have the same name so we can use either one.
+                alias: field_b.field.name.clone(),
+                original_location: field_a.field.location(),
+                original_coordinate: field_a.coordinate().with_argument(arg.name.clone()),
+                original_value: original_arg.map(|arg| (*arg.value).clone()),
+                conflicting_location: field_b.field.location(),
+                conflicting_coordinate: field_b.coordinate().with_argument(arg.name.clone()),
+                conflicting_value: redefined_arg.map(|arg| (*arg.value).clone()),
+            }
+        };
+
+    // Check if fieldB provides the same argument names and values as fieldA (order-independent).
+    for arg in &field_a.field.arguments {
+        let Some(other_arg) = field_b.field.argument_by_name(&arg.name) else {
```

Comment:
I don’t expect this to ever be a problem in non-adversarial documents, but this takes O(n*m) time. Maybe convert both argument lists to maps if either has more than like 20 entries?

Reply:
I added an `ArgumentLookup` type for this, it's a good shout. It could also potentially happen in arguments validation when matching values to the argument definition, however there `m` is decided by the schema, so `O(m*n)` doesn't grow exponentially.